File: NYUDining/RFDiningLocation.m
//
//  DiningLocation.m
//  NYUDining
//
//  Created by Ross Freeman on 9/8/15.
//  Copyright (c) 2015 Ross Freeman. All rights reserved.
//

#import "RFDiningLocation.h"

@implementation RFDiningLocation

-(id)initWithData:(NSDictionary *)data andParameters:(NSDictionary *)params{
    self = [super init];
    
    if (self) {
        _hours = [[NSMutableArray alloc] initWithCapacity:0];
        _paymentTypes = [[NSMutableArray alloc] initWithCapacity:0];
        _coordinates = [[NSMutableArray alloc] initWithCapacity:0];
        self.data = data;
        
        NSString *cal = [params objectForKey:@"calendar"];
        
        _hours = [data objectForKey:cal];
        _name = [data objectForKey:@"Name"];
        _logoURL = [data objectForKey:@"logo_URL"];
        _menuURL = [data objectForKey:@"menu_url"];
        _address = [data objectForKey:@"Address"];
        _phoneNumber = [data objectForKey:@"phone"];
        _coordinates = [[data objectForKey:@"Coordinates"] mutableCopy];
        
        for (NSDictionary *payments in [data objectForKey:@"payments"]) {
            if ([payments objectForKey:@"accepted"]) {
                [_paymentTypes addObject:[payments objectForKey:@"name"]];
            }
        }
        
    }
    
    return self;
}

// Changes hours selection
-(void)setNewHours:(NSString *)option {
    
    option = [option stringByReplacingOccurrencesOfString:@" " withString:@"_"];
    
    _hours = [self.data objectForKey:option];
    
    // Recalculate status based on new times
    [self isOpen];
}


// Determines if location is currently open
-(BOOL)isOpen {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"EEEE"];
    NSString *dayOfWeek = [dateFormatter stringFromDate:[NSDate date]];
    
    // Finds the current time and stores it as a double
    NSDate *now = [NSDate date];
    NSDateFormatter *outputFormatter = [[NSDateFormatter alloc] init];
    [outputFormatter setDateFormat:@"HH:mm"];
    NSString *currentTime = [outputFormatter stringFromDate:now];
    NSArray *times = [currentTime componentsSeparatedByString:@":"];
    double currentHour = [times[0] doubleValue];
    double currentMinute = [times[1] doubleValue];
    NSLog(@"Current time: %ld:%ld", (long)currentHour, (long)currentMinute);
    double timeDouble = currentHour + currentMinute/60;
    
    double timeA = 0;
    double timeB = 0;

    [outputFormatter setDateFormat:@"a"];
    
    NSString *hoursToday = @"";
    
    // Get hours for today
    if ([dayOfWeek isEqualToString:@"Sunday"]) {
        hoursToday = _hours[0];
    }
    
    else if ([dayOfWeek isEqualToString:@"Monday"]) {
        hoursToday = _hours[1];
    }
    
    else if ([dayOfWeek isEqualToString:@"Tuesday"]) {
        hoursToday = _hours[2];
    }
    
    else if ([dayOfWeek isEqualToString:@"Wednesday"]) {
        hoursToday = _hours[3];
    }
    
    else if ([dayOfWeek isEqualToString:@"Thursday"]) {
        hoursToday = _hours[4];
    }
    
    else if ([dayOfWeek isEqualToString:@"Friday"]) {
        hoursToday = _hours[5];
    }
    
    else if ([dayOfWeek isEqualToString:@"Saturday"]) {
        hoursToday = _hours[6];
    }
    
    // Check if location closes mid-day
    if ([hoursToday containsString:@","]) {
        NSArray *timeSlots = [hoursToday componentsSeparatedByString:@","];
        
        NSInteger counter = 0;
        
        for (NSString *timeSlot in timeSlots) {
            
            counter++;
            
            NSArray *times = [timeSlot componentsSeparatedByString:@"-"];
            
            NSMutableString *time0 = [times[0] mutableCopy];
            NSMutableString *time1 = [times[1] mutableCopy];
            
            if ([times[0] containsString:@"am"]) {
                
                if ([times[0] containsString:@"12"]) {
                    timeA = 0;
                }
                
                else {
                    [time0 deleteCharactersInRange:[time0 rangeOfString:@"am"]];
                    
                    if ([time0 containsString:@":"]) {
                        NSArray *timeArray = [time0 componentsSeparatedByString:@":"];
                        NSString *minute = timeArray[1];
                        double minuteInHour = [minute doubleValue] / 60.0;
                        timeA = [times[0] doubleValue] + minuteInHour;
                    }
                    
                    else
                        timeA = [times[0] doubleValue];
                }
            }
            
            else {
                [time0 deleteCharactersInRange:[time0 rangeOfString:@"pm"]];
                
                
                if ([time0 containsString:@":"]) {
                    NSArray *timeArray = [time0 componentsSeparatedByString:@":"];
                    NSString *minute = timeArray[1];
                    double minuteInHour = [minute doubleValue] / 60.0;
                    timeA = [times[0] doubleValue] + minuteInHour;
                }
                
                else
                    timeA = [times[0] doubleValue];
                
                if (timeA < 12) {
                    timeA += 12;
                }
            }
            
            
            
            if ([times[1] containsString:@"am"]) {
                if ([times[1] containsString:@"12"]) {
                    timeB = 0;
                }
                
                else {
                    [time1 deleteCharactersInRange:[time1 rangeOfString:@"am"]];
                    
                    if ([time1 containsString:@":"]) {
                        NSArray *timeArray = [time1 componentsSeparatedByString:@":"];
                        NSString *minute = timeArray[1];
                        double minuteInHour = [minute doubleValue] / 60.0;
                        timeB = [times[1] doubleValue] + minuteInHour;
                    }
                    
                    else
                        timeB = [times[1] doubleValue];
                }
            }
            
            else {
                [time1 deleteCharactersInRange:[time1 rangeOfString:@"pm"]];
                
                
                if ([time1 containsString:@":"]) {
                    timeB = [times[1] doubleValue] + 0.5;
                }
                
                else
                    timeB = [times[1] doubleValue];
                
                if (timeB < 12) {
                    timeB += 12;
                }
            }
            
            if (timeDouble >= timeA && timeDouble < timeB) {
                return YES;
            }
            
            else if (counter == timeSlots.count)
                return NO;
            
        }
        
        
    }
    
    else {
        
        if ([hoursToday isEqualToString:@"Closed"]) {
            return NO;
        }
        else {
            
            NSArray *times = [hoursToday componentsSeparatedByString:@"-"];
            
            NSMutableString *time0 = [times[0] mutableCopy];
            NSMutableString *time1 = [times[1] mutableCopy];
            
            if ([times[0] containsString:@"am"]) {
                
                if ([times[0] containsString:@"12"]) {
                    timeA = 0;
                }
                
                else {
                    [time0 deleteCharactersInRange:[time0 rangeOfString:@"am"]];
                    
                    if ([time0 containsString:@":"]) {
                        NSArray *timeArray = [time0 componentsSeparatedByString:@":"];
                        NSString *minute = timeArray[1];
                        double minuteInHour = [minute doubleValue] / 60.0;
                        timeA = [times[0] doubleValue] + minuteInHour;
                    }
                    
                    else
                        timeA = [times[0] doubleValue];
                }
            }
            
            else {
                [time0 deleteCharactersInRange:[time0 rangeOfString:@"pm"]];
                
                if ([time0 containsString:@":"]) {
                    NSArray *timeArray = [time0 componentsSeparatedByString:@":"];
                    NSString *minute = timeArray[1];
                    double minuteInHour = [minute doubleValue] / 60.0;
                    timeA = [times[0] doubleValue] + minuteInHour;
                }
                
                else
                    timeA = [times[0] doubleValue];
                
                if (timeA < 12) {
                    timeA += 12;
                }
            }
            
            if ([times[1] containsString:@"am"]) {
                if ([times[1] containsString:@"12"]) {
                    timeB = 0;
                }
                
                else {
                    [time1 deleteCharactersInRange:[time1 rangeOfString:@"am"]];
                    
                    if ([time1 containsString:@":"]) {
                        NSArray *timeArray = [time1 componentsSeparatedByString:@":"];
                        NSString *minute = timeArray[1];
                        double minuteInHour = [minute doubleValue] / 60.0;
                        timeB = [times[1] doubleValue] + minuteInHour;
                    }
                    
                    else
                        timeB = [times[1] doubleValue];
                }
            }
            
            else {
                [time1 deleteCharactersInRange:[time1 rangeOfString:@"pm"]];
                
                if ([time1 containsString:@":"]) {
                    NSArray *timeArray = [time1 componentsSeparatedByString:@":"];
                    NSString *minute = timeArray[1];
                    double minuteInHour = [minute doubleValue] / 60.0;
                    timeB = [times[1] doubleValue] + minuteInHour;
                }
                
                else
                    timeB = [times[1] doubleValue];
                
                if (timeB < 12) {
                    timeB += 12;
                }
            }
            
            if ((timeDouble >= timeA && timeDouble < timeB) || (timeDouble >= timeA && timeB == 0)) {
                return YES;
            }
            
            else
                return NO;
            
        }
    }
    
    return YES;
}

// For dev purposes ONLY
// Do not use
- (BOOL)isOpenTest {
    return NO;
}

- (NSDate *)convertTimeRangeToDate:(NSString *)time {
    
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    NSDate *formattedTime = [formatter dateFromString:time];
    
    NSLog(@"%@", formattedTime);
    
    return formattedTime;
    
}


@end

File: NYUDining/RFDiningLocation.swift
//
//  RFDiningLocation.swift
//  
//
//  Created by Ross Freeman on 6/10/16.
//
//

import Foundation

enum DayOfWeek: String {
    case Sunday = "Sunday"
    case Monday = "Monday"
    case Tuesday = "Tuesday"
    case Wednesday = "Wednesday"
    case Thursday = "Thursday"
    case Friday = "Friday"
    case Saturday = "Saturday"
    case Closed = "Closed"
}

class RFDiningLocation: NSObject {
    
    let name: String!
    let logoURL: String!
    let menuURL: String?
    let hours: [String]!
    let address: String!
    let coordinates: [Double]
    let data: [String: AnyObject]
    
    init(data: [String: AnyObject], params: [String: AnyObject]) {
        self.data = data
        
        let cal: String = params["calendar"] as! String
        hours = data[cal] as! [String]
        name = data["Name"] as! String
        logoURL = data["logo_URL"] as! String
        address = data["Address"] as! String
        menuURL = data["menu_url"] as? String
        coordinates = data["Coordinates"] as! [Double]
        
    }
     
    func isOpen() -> Bool {
        let now = NSDate()
        let dateFormatter = NSDateFormatter()
        dateFormatter.dateFormat = "EEEE"
        let dayOfWeek = DayOfWeek(rawValue: dateFormatter.stringFromDate(now))
        
        var hoursToday = ""
        
        if let dayOfWeek = dayOfWeek {
            switch dayOfWeek {
            case .Sunday:
                hoursToday = hours[0]
            case .Monday:
                hoursToday = hours[1]
            case .Tuesday:
                hoursToday = hours[2]
            case .Wednesday:
                hoursToday = hours[3]
            case .Thursday:
                hoursToday = hours[4]
            case .Friday:
                hoursToday = hours[5]
            default:
                hoursToday = hours[6]
            }
        }
        
        if hoursToday.containsString(",") {
            let timeComponents = hoursToday.componentsSeparatedByString(",")
            return isTimeInRange(timeComponents[0]) || isTimeInRange(timeComponents[1])
        }
        
        else {
            return isTimeInRange(hoursToday)
        }
    }
    
    private func normalizedTime() -> NSDate {
        let now = NSDate()
        
        let dateFormatter = NSDateFormatter()
        dateFormatter.timeStyle = .ShortStyle
        
        let timeStr = dateFormatter.stringFromDate(now)
        return dateFormatter.dateFromString(timeStr)!
        
    }
    
    private func isTimeInRange(timeRange: String) -> Bool {
        
        if timeRange == "Closed" {
            return false
        }
        
        let calendar = NSCalendar.currentCalendar()
        let now = normalizedTime()
        
        let dateFormatter = NSDateFormatter()
        dateFormatter.timeStyle = .ShortStyle
        
        let timeComponents = timeRange.componentsSeparatedByString("-")
        
        let openTime = dateFormatter.dateFromString(timeComponents[0])
        let closeTime = dateFormatter.dateFromString(timeComponents[1])
        
        if let openTime = openTime, var closeTime = closeTime {
            
            // Check if close time is 12am
            let diffComparison = calendar.compareDate(openTime, toDate: closeTime, toUnitGranularity: .Hour)
            if diffComparison == .OrderedDescending {
                guard let newTime = calendar.dateByAddingUnit(.Day, value: 1, toDate: closeTime, options: []) else {
                    return false
                }
                closeTime = newTime
            }
            
            let openComparison = calendar.compareDate(now, toDate: openTime, toUnitGranularity: .Minute)
            let closeComparison = calendar.compareDate(now, toDate: closeTime, toUnitGranularity: .Minute)
            
            return (openComparison == .OrderedDescending && closeComparison == .OrderedAscending)
        }
        
        return false
    }
    
}